File: src/features/plugin/armazenamento.py
import json
import os
import stat
import tempfile
from collections.abc import Callable, MutableMapping
from pathlib import Path
from typing import Literal, cast

import json5
import tomlkit

from src.features.plugin.modelos import ConfigFormatError

FormatName = Literal["json", "json5", "toml"]
MappingValue = MutableMapping[str, object]


def load_document(path: Path, format_name: FormatName) -> dict[str, object]:
    try:
        content = path.read_text(encoding="utf-8")
        if format_name == "json":
            document = json.loads(content)
        elif format_name == "json5":
            document = json5.loads(content)
        elif format_name == "toml":
            document = tomlkit.parse(content)
        else:
            raise ConfigFormatError(f"unsupported configuration format: {format_name}")
    except ConfigFormatError:
        raise
    except (OSError, ValueError, TypeError) as error:
        raise ConfigFormatError(f"invalid {format_name} configuration: {path}") from error

    if not isinstance(document, dict):
        raise ConfigFormatError(f"configuration root must be a mapping: {path}")
    return cast(dict[str, object], document)


def _find_container(
    document: dict[str, object], container_path: tuple[str, ...], create_missing: bool
) -> MappingValue | None:
    current = cast(MappingValue, document)
    for key in container_path:
        if key not in current:
            if not create_missing:
                return None
            current[key] = {}
        value = current[key]
        if isinstance(value, MutableMapping):
            current = cast(MappingValue, value)
        else:
            return None
    return current


def merge_mcp_entry(
    document: dict[str, object],
    container_path: tuple[str, ...],
    name: str,
    entry: dict[str, object],
    owns: Callable[[object], bool],
) -> tuple[dict[str, object], bool]:
    container = _find_container(document, container_path, create_missing=True)
    if container is None:
        return document, False

    current = container.get(name)
    if name not in container:
        container[name] = entry
        return document, True
    if owns(current):
        if current == entry:
            return document, False
        container[name] = entry
        return document, True
    return document, False


def remove_mcp_entry(
    document: dict[str, object],
    container_path: tuple[str, ...],
    name: str,
    owns: Callable[[object], bool],
) -> tuple[dict[str, object], bool]:
    container = _find_container(document, container_path, create_missing=False)
    if container is None or name not in container:
        return document, False
    if not owns(container[name]):
        return document, False
    del container[name]
    return document, True


def dump_document(document: dict[str, object], format_name: FormatName) -> str:
    if format_name == "json":
        return f"{json.dumps(document, indent=2)}\n"
    if format_name == "json5":
        return f"{json5.dumps(document, indent=2)}\n"
    if format_name == "toml":
        return tomlkit.dumps(document)
    raise ConfigFormatError(f"unsupported configuration format: {format_name}")


def atomic_write(path: Path, content: str) -> None:
    temporary_path: Path | None = None
    existing_mode: int | None = None
    try:
        try:
            existing_mode = stat.S_IMODE(path.stat().st_mode)
        except FileNotFoundError:
            pass

        with tempfile.NamedTemporaryFile(
            mode="w",
            encoding="utf-8",
            dir=path.parent,
            prefix=f".{path.name}.",
            suffix=".tmp",
            delete=False,
        ) as temporary:
            temporary_path = Path(temporary.name)
            if existing_mode is not None:
                os.chmod(temporary_path, existing_mode)
            temporary.write(content)
            temporary.flush()
            os.fsync(temporary.fileno())

        os.replace(temporary_path, path)
    finally:
        if temporary_path is not None:
            temporary_path.unlink(missing_ok=True)

File: src/features/plugin/armazenamento.ts
import { chmodSync, readFileSync, renameSync, rmSync, statSync, writeFileSync } from "node:fs";
import { basename, dirname, join } from "node:path";

import json5 from "json5";
import { parse as parseToml, stringify as stringifyToml } from "smol-toml";

import { ConfigFormatError } from "./modelo.js";

export type FormatName = "json" | "json5" | "toml";

type Mapping = Record<string, unknown>;

export function loadDocument(path: string, formatName: FormatName): Mapping {
  let document: unknown;
  try {
    const content = readFileSync(path, "utf-8");
    if (formatName === "json") {
      document = JSON.parse(content);
    } else if (formatName === "json5") {
      document = json5.parse(content);
    } else if (formatName === "toml") {
      document = parseToml(content);
    } else {
      throw new ConfigFormatError(`unsupported configuration format: ${formatName}`);
    }
  } catch (error) {
    if (error instanceof ConfigFormatError) {
      throw error;
    }
    throw new ConfigFormatError(`invalid ${formatName} configuration: ${path}`);
  }

  if (typeof document !== "object" || document === null || Array.isArray(document)) {
    throw new ConfigFormatError(`configuration root must be a mapping: ${path}`);
  }
  return document as Mapping;
}

function deepEquals(a: unknown, b: unknown): boolean {
  if (a === b) {
    return true;
  }
  if (Array.isArray(a) && Array.isArray(b)) {
    return a.length === b.length && a.every((item, index) => deepEquals(item, b[index]));
  }
  if (
    typeof a === "object" && a !== null && !Array.isArray(a) &&
    typeof b === "object" && b !== null && !Array.isArray(b)
  ) {
    const left = a as Mapping;
    const right = b as Mapping;
    const leftKeys = Object.keys(left);
    return (
      leftKeys.length === Object.keys(right).length &&
      leftKeys.every((key) => key in right && deepEquals(left[key], right[key]))
    );
  }
  return false;
}

function findContainer(
  document: Mapping,
  containerPath: readonly string[],
  createMissing: boolean,
): Mapping | null {
  let current: Mapping = document;
  for (const key of containerPath) {
    if (!(key in current)) {
      if (!createMissing) {
        return null;
      }
      current[key] = {};
    }
    const value = current[key];
    if (typeof value === "object" && value !== null && !Array.isArray(value)) {
      current = value as Mapping;
    } else {
      return null;
    }
  }
  return current;
}

export function mergeMcpEntry(
  document: Mapping,
  containerPath: readonly string[],
  name: string,
  entry: Mapping,
  owns: (value: unknown) => boolean,
): [Mapping, boolean] {
  const container = findContainer(document, containerPath, true);
  if (container === null) {
    return [document, false];
  }

  const current = container[name];
  if (!(name in container)) {
    container[name] = entry;
    return [document, true];
  }
  if (owns(current)) {
    if (deepEquals(current, entry)) {
      return [document, false];
    }
    container[name] = entry;
    return [document, true];
  }
  return [document, false];
}

export function removeMcpEntry(
  document: Mapping,
  containerPath: readonly string[],
  name: string,
  owns: (value: unknown) => boolean,
): [Mapping, boolean] {
  const container = findContainer(document, containerPath, false);
  if (container === null || !(name in container)) {
    return [document, false];
  }
  if (!owns(container[name])) {
    return [document, false];
  }
  delete container[name];
  return [document, true];
}

export function dumpDocument(document: Mapping, formatName: FormatName): string {
  if (formatName === "json") {
    return `${JSON.stringify(document, null, 2)}\n`;
  }
  if (formatName === "json5") {
    return `${json5.stringify(document, null, 2)}\n`;
  }
  if (formatName === "toml") {
    return stringifyToml(document);
  }
  throw new ConfigFormatError(`unsupported configuration format: ${formatName}`);
}

export function atomicWrite(path: string, content: string): void {
  // ponytail: no fsync; single-user CLI config files, renameSync is already
  // atomic on the same volume.
  const temporaryPath = join(
    dirname(path),
    `.${basename(path)}.${process.pid}.${Math.random().toString(36).slice(2)}.tmp`,
  );
  try {
    let existingMode: number | null = null;
    try {
      existingMode = statSync(path).mode & 0o777;
    } catch {
      // FileNotFoundError equivalent: file does not exist yet
    }

    writeFileSync(temporaryPath, content, "utf-8");
    if (existingMode !== null) {
      chmodSync(temporaryPath, existingMode);
    }
    renameSync(temporaryPath, path);
  } finally {
    rmSync(temporaryPath, { force: true });
  }
}
